sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImpl.java

```diff
@@ -210,6 +212,224 @@ public Row getCurrentRow() throws NoSuchElementException {
+      SettableFuture<UpstreamResults> f = SettableFuture.create();
+      // When the nextRequest is null, the last fill completed and the buffer contains the last rows
+      if (nextRequest == null) {
+        f.set(new UpstreamResults(ImmutableList.of(), null));
+        return f;
+      }
+
+      // TODO(diegomez): Remove atomic ScanHandler for simpler StreamObserver/Future implementation
+      AtomicReference<ScanHandler> atomicScanHandler = new AtomicReference<>();
+      ScanHandler handler =
+          session
+              .getDataClient()
+              .readFlatRows(
+                  nextRequest,
+                  new StreamObserver<FlatRow>() {
+                    List<Row> rows = new ArrayList<>();
+                    long currentByteSize = 0;
+                    boolean byteLimitReached = false;
+
+                    @Override
+                    public void onNext(FlatRow flatRow) {
+                      Row row = FlatRowConverter.convert(flatRow);
+                      currentByteSize += row.getSerializedSize();
+                      rows.add(row);
+
+                      if (currentByteSize > maxSegmentByteSize) {
+                        byteLimitReached = true;
+                        atomicScanHandler.get().cancel();
+                        return;
+                      }
+                    }
+
+                    @Override
+                    public void onError(Throwable e) {
+                      f.setException(e);
+                    }
+
+                    @Override
+                    public void onCompleted() {
+                      ReadRowsRequest nextNextRequest = null;
+
+                      // When requested rows < limit, the current request will be the last
+                      if (byteLimitReached || rows.size() == nextRequest.getRowsLimit()) {
+                        nextNextRequest =
+                            truncateRequest(nextRequest, rows.get(rows.size() - 1).getKey());
+                      }
+                      f.set(new UpstreamResults(rows, nextNextRequest));
+                    }
+                  });
+      atomicScanHandler.set(handler);
+      return f;
+    }
+
+    private void waitReadRowsFuture() throws IOException {
+      try {
+        UpstreamResults r = future.get();
+        buffer.addAll(r.rows);
+        nextRequest = r.nextRequest;
+        future = null;
+        serviceCallMetric.call("ok");
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new IOException(e);
+      } catch (ExecutionException e) {
+        throw new IOException(e.getCause());
+      }
+      catch (Exception e) {
+        if (e instanceof InterruptedException) {
+          Thread.currentThread().interrupt();
+          throw new IOException(e);
+        }
+        if (e.getCause() instanceof StatusRuntimeException) {
+          throw new IOException(e);
+        }
+      }
```

Comment:
This looks unfinished:
- you are checking InterruptedException twice
- you lost the metric increment

I think you want something like:
```
try {
...
} catch (InterruptedException e) {
  Thread.currentThread().interrupt();
  throw new IOException(e);
} catch(ExecutionException e) {
  Throwable cause = e.getCause();
  if (cause instanceof StatusRuntimeException) {
    // increment error counts
  }
  throw new IOException(cause);
} 
```